lib/ecto/changeset.ex

```diff
@@ -1372,7 +1372,10 @@ defmodule Ecto.Changeset do
 
     case new_errors do
       [] -> %{changeset | required: fields ++ required}
-      _  -> %{changeset | required: fields ++ required, errors: new_errors ++ errors, valid?: false}
+      _  ->
+        new_errors_keys = Keyword.keys(new_errors)
```

Comment:
What if we change the `for` that computes `new_errors` above to simply return `fields_with_errors` and then we build the messages by doing another pass?

Another idea is to use a reduce and build the errors and changes at once.